Webpack powered Hotwire Storefront

damianlegawiec opened the pull request:
Fixes https://github.com/spree/spree/issues/8789 
Fixes https://github.com/spree/spree/issues/10850
Fixes https://github.com/spree/spree/issues/10853
Fixes https://github.com/spree/spree/issues/10849

damianlegawiec committed:
Added webpacker dependency

damianlegawiec committed:
Added basic NPM package skeleton with microbundle

damianlegawiec committed:
Added Stimulus application skeleton

damianlegawiec committed:
Added CouponCode Stimulus controller

tniezg commented on frontend/package.json:
```diff
@@ -0,0 +1,23 @@
+  "module": "dist/spree_frontend.module.js",
+  "unpkg": "dist/spree_frontend.umd.js",
+  "scripts": {
+    "build": "microbundle",
```
```suggestion
    "build": "microbundle --globals @spree/storefront-api-v2-sdk=SpreeSDK",
```

tniezg commented on frontend/package.json:
```diff
@@ -0,0 +1,23 @@
+  "source": "app/javascript/index.js",
+  "main": "dist/spree_frontend.js",
+  "exports": "dist/spree_frontend.modern.js",
+  "module": "dist/spree_frontend.module.js",
```
Webpack can't generate ES6 modules: https://github.com/webpack/webpack/issues/2933#issuecomment-774253975. For `frontend/dist/spree_frontend.module.js` to work, the Spree SDK would have to use Rollup instead of Webpack. `dist/server/index.js` and `dist/client/index.js` are exported in the UMD format. Most importantly, these exports support importing the SpreeSDK via CommonJS in Node and in the browser and referencing it by `window.SpreeSDK` in the browser.

tniezg replied:
https://github.com/spree/spree/pull/10982

damianlegawiec committed:
Update frontend/package.json

Co-authored-by: Tomek Niezgoda <1410097+tniezg@users.noreply.github.com>

damianlegawiec committed:
Setup Storefront API <> Frontend integration

damianlegawiec committed:
Working Coupon Code controller :)

tniezg commented on frontend/app/javascript/controllers/coupon_code_controller.js:
```diff
@@ -0,0 +1,67 @@
+
+  handleResponse(response) {
+    if (response.isFail()) {
+      this.handleError(response.fail().serverResponse.data.error)
```
To handle all Spree responses, including unexpected ones, a check like `response.fail() instanceof window.SpreeSDK.errors.BasicSpreeError` is needed here. If `response.fail()` is an instance of `BasicSpreeError` or its subclass, then `response.fail().summary` is the same as `response.fail().serverResponse.data.error`. If `response.fail()` is a superclass of `BasicSpreeError` it means there was a more serious error.

Commit:
Relative module paths in package.json

Commit:
Do not generate broken ES6 module and share script config

Commit:
Use CommonJS server-side Spree SDK

Commit:
Refresh bundles

Commit:
Remove broken ES6 bundles

damianlegawiec committed:
Fix test suite with webpacker

Commit:
Remove redundant SDK import

Commit:
Remove microbundle generated files

Commit:
Switch to Webpack

Commit:
Replace microbundle bundles with Webpack bundles

Commit:
Remove package-lock.json and update yarn.lock

Commit:
Refresh UMD bundles

Commit:
Use yarn to call build script from dev script

Commit:
Load all Stimulus JS controllers inside controllers/

damianlegawiec committed:
Setup test app with webpacker and frontend NPM package

damianlegawiec committed:
Added JS log debugging to feature specs

damianlegawiec committed:
Added CheckoutHelper with checkout_submit_tag method

damianlegawiec committed:
Fixed passing Storefront API host

damianlegawiec committed:
Added new coupon code applier to Checkout

damianlegawiec committed:
Revert "Added JS log debugging to feature specs"

This reverts commit e45dd0915d0152b0c5cd313dbe3640fbbc84f161.

damianlegawiec committed:
Don't recompile webpacker in test env

tniezg commented:
Stimulurs controllers export from @spree/frontend: https://github.com/spree/spree/pull/10991

damianlegawiec committed:
Remove dist files from repo

damianlegawiec committed:
Rename @spree/frontend to @spree/storefront

damianlegawiec committed:
Replace turbolinks with turbo drive

damianlegawiec committed:
Extract methods for Cart management into Spree::CartMethods module

damianlegawiec committed:
Setup Turbo / Stimulus application

damianlegawiec committed:
Remove jquery-based JS code

This will be extracted to a seperate gem to ease the upgrade process

damianlegawiec committed:
Support flashes with turbo frames/streams

damianlegawiec committed:
Rebuilt nav bar with turbo frames / stimulus

damianlegawiec committed:
Rebuilt product carousels with turbo frames

damianlegawiec committed:
New Cart page powered by Turbo

damianlegawiec committed:
Run only specs from ./spec/ directory

damianlegawiec committed:
Don't precompile assets for API, Core and Sample gems

damianlegawiec committed:
Generate Procfile for Sandbox

damianlegawiec committed:
Fix deleting addresses Turbo way

damianlegawiec committed:
Compile webpack only once per test suite

damianlegawiec committed:
Fixed cart layout

damianlegawiec committed:
Don't try to enforce authorization when order does not exist

damianlegawiec committed:
Fixed turbo cart preference toggle

damianlegawiec committed:
Remove API host preference

damianlegawiec committed:
Include CacheHelper for all Storefront controllers

damianlegawiec committed:
Remove Spree Multi Domain from Sandbox

This gem is not compatible with turbo/hotwire

damianlegawiec committed:
Move Ruby code for Qty component to Cart Helper

damianlegawiec committed:
Fixed Currency / Locale dropdown with Turbo

damianlegawiec committed:
Fixed desktop mega nav

damianlegawiec committed:
Product list Mobile UI

damianlegawiec committed:
Fixed transitioning on some pages / forms with Turbo

In turbolinks we've used `method: :get` to mark which actions/pages should not be cached in Turbolinks and which require full reload. This was changes in Turbo to `data-turbo="false"`. Some pages such as checkout do not require turning off Turbolinks anymore since https://github.com/spree/spree/commit/175aa687287800e1ee7dd1536a7e32c36df5097b
